### PR TITLE
Worldgen2

### PR DESCRIPTION
**Pull request by koskja**
# Reimplementing world generation

## Status

- [x] Ready 
- [ ] Development
- [ ] Hold

## Description

Added a new WorldSource that generates new chunks, updated chunk biomes for 1.16, redid flat world gen and tweaked default worldgen to a more normal state.

## Related issues

World generation #24 

## Checklist

- [x] Ran `cargo fmt`, `cargo clippy`, `cargo build --release` and `cargo test` and fixed any generated errors!
- [x] Removed unnecessary commented out code
- [ ] Used specific traces (if you trace actions please specify the cause i.e. the player)

### CODE REVIEW

File: feather/common/src/world_source/generating.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+impl GeneratingWorldSource {
+    pub fn from_arc<G>(generator: Arc<G>) -> Self
+    where
+        G: WorldGenerator + 'static,
+    {
+        let (send_pos, recv_pos) = unbounded::<ChunkPosition>();
+        let (send_gen, recv_gen) = unbounded::<LoadedChunk>();
+        for _ in 0..8 {
+            // FIXME: do not hardcode '8' as a magic number
```
Comment (caelunshun) [inline feather/common/src/world_source/generating.rs:28]: I'd recommend starting just one thread here that polls the channel, then invokes `rayon::spawn` for each chunk that needs to be generated. Right now, we're blocking the Rayon thread poll by waiting on channels, which can cause a deadlock

File: feather/server/config.toml [hunk 2]
```diff
@@ -12,7 +12,7 @@ online_mode = true
 motd = "A Feather server"
 max_players = 16
 default_gamemode = "creative"
-view_distance = 12
+view_distance = 32
```
Comment (caelunshun) [inline feather/server/config.toml:15]: This should probably be reverted? 12 is the default for vanilla too.

Comment (koskja) [reply feather/server/config.toml:15]: I thought that the global threadpool can have an unlimited amount of threads? Also, while the channels are blocked, other threads can work in the meantime. And of course, the view_distance should be set to 12.

Comment (caelunshun) [reply feather/server/config.toml:15]: > I thought that the global threadpool can have an unlimited amount of threads?

It's a fixed size AFAIK. See https://docs.rs/rayon-core/1.9.1/rayon_core/struct.ThreadPoolBuilder.html#method.num_threads

> Also, while the channels are blocked, other threads can work in the meantime.

Yes, but the Rayon threads are blocked. Which means other tasks on the Rayon thread pool can't run.

Comment (koskja) [reply feather/server/config.toml:15]: Just learned that rayon threads are not the same as os threads :) I am not a fan of making a separate thread for every chunk, do you think a 'personal' ThreadPool would be a good idea?

Comment (caelunshun) [reply feather/server/config.toml:15]: Using `rayon::spawn` won't create a new thread for each chunk. It dispatches a task into a pool of existing threads, which is very cheap.

**Comment by koskja** [comment]
review?

**Review by caelunshun — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**